src/traces/table/render.js

```diff
@@ -0,0 +1,561 @@
+/**
+* Copyright 2012-2017, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var c = require('./constants');
+var d3 = require('d3');
+var Drawing = require('../../components/drawing');
+
+function keyFun(d) {return d.key;}
+
+function repeat(d) {return [d];}
+
+function unwrap(d) {
+    return d[0]; // plotly data structure convention
+}
+
+function model(layout, d, i) {
+    var cd0 = unwrap(d),
+        trace = cd0.trace,
+        domain = trace.domain,
+        width = layout.width,
+        labels = trace.labels,
+        columnWidths = trace.columnwidth;
+
+    var colCount = labels.length;
+
+    var groupWidth = Math.floor(width * (domain.x[1] - domain.x[0]));
+    var groupHeight = Math.floor(layout.height * (domain.y[1] - domain.y[0]));
+
+    columnWidths = trace.header.values.map(function(d, i) {
+        return Array.isArray(columnWidths) ?
+            columnWidths[Math.min(i, columnWidths.length - 1)] :
+            isFinite(columnWidths) && columnWidths !== null ? columnWidths : 1;
+    });
+
+    var totalColumnWidths = columnWidths.reduce(function(p, n) {return p + n;}, 0);
+    columnWidths = columnWidths.map(function(d) {return d / totalColumnWidths * groupWidth;});
+
+    var pad = layout.margin || {l: 80, r: 80, t: 100, b: 80};
```

Comment:
you shouldn't need that fallback. `layout.margin` gets filled in during supply defaults.